presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonFunctions.java

```diff
@@ -71,6 +77,101 @@
 
     private static final ObjectMapper SORTED_MAPPER = new JsonObjectMapperProvider().get().configure(ORDER_MAP_ENTRIES_BY_KEYS, true);
 
+    private static final ParseContext jaywayParseCtx = com.jayway.jsonpath.JsonPath.using(com.jayway.jsonpath.Configuration.builder().jsonProvider(
+            new JacksonJsonNodeJsonProvider()).build());
+
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    private static JsonNode jaywayExtract(Slice json, JsonPath jsonPath) {
+        try {
+            if (jsonPath.getPattern().isEmpty()) {
+                // for some reason, jayway throws IllegalArgumentException for an empty path, but an InvalidPathException for other invalid paths
+                throw new InvalidPathException();
+            }
+            Object res = jaywayParseCtx.parse(json.getInput()).read(jsonPath.getPattern());
+            if (res instanceof JsonNode) {
+                return (JsonNode) res;
+            } else {
+                // Jayway will respect Jackson mappings as provided in the configuration and return a JsonNode for simple cases.
+                // But for JsonPath functions ($.avg, ...), it will return a Java boxed type (Double, String etc.) instead
+                // of a properly formed JsonNode. This is why we need to re-create a JsonNode in that case
+                return mapper.valueToTree(res);
+            }
+        } catch (InvalidJsonException ex) {
+            // replicate Presto's JsonPath behaviour: if the input JSON is invalid, then return NULL
+            // instead of throwing an exception
+            return null;
+        } catch (InvalidPathException ex) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Invalid JSON path: '%s'", jsonPath.getPattern(), ex.getMessage()));
+        }
+    }
+
+    private static Slice jsonPathExtract(Slice json, JsonPath jsonPath, JsonPathEngine engine) {
+        switch (engine) {
+            case PRESTO:
+                return JsonExtract.extract(json, jsonPath.getObjectExtractor());
+            case JAYWAY:
+                JsonNode node = jaywayExtract(json, jsonPath);
+                if (node == null) {
+                    return null;
+                }
+                return utf8Slice(node.toString());
+            default:
```

Comment:
I think it's doable, but that means we'd be slowing down simple JsonPath expressions that don't need Jayway and that are fast today, since we'd need to run a first pass on the jsonpath pattern.